agenthub/README.md

```diff
@@ -26,13 +26,15 @@ The `state` contains:
 Here is a list of available Actions, which can be returned by `agent.step()`:
 - [`CmdRunAction`](../opendevin/action/bash.py) - Runs a command inside a sandboxed terminal
 - [`CmdKillAction`](../opendevin/action/bash.py) - Kills a background command
+- [`IPythonRunCellAction`](../opendevin/action/bash.py) - Execute a block of Python code interactively (in Jupyter notebook) and receives `CmdOutputObservation`. Requires setting up `jupyter` [plugin](../opendevin/sandbox/plugins) as a requirement.
```

Comment:
Can we make this what we called a "pseudo-action" or "auxiliiary action" in your other PR? i.e. keep it internal to CodeAct, and just split it into Write and Run actions under the hood?

Reply:
I'm still somewhat debating this -- I guess it ultimately comes down to **whether we want to visualize Jupyter code execution in our frontend**. If so, keeping them might be valuable.

You can see the original CodeAct video for one use case I'm imagining: **plotting figures!** This is already support with the current implementations, we just need to visualize the cell output in our front-end.

These figures are images encoded in `base64` from Jupyter. If we did not do any special handling for Jupyter (i.e., `IPythonRunCellAction` and `IPythonRunCellObservation`), these `base64` images would be interpreted as CmdOutput and get printed to the terminal output, which can be extremely messy.

<video src="https://github.com/xingyaoww/code-act/assets/38853559/62c80ada-62ce-447e-811c-fc801dd4beac"> </video>


If our initial MVP baseline is a ChatGPT-style Python interpreter + Bash command line, I think it could make sense for us to incorporate `IPython Jupyter` as a core action directly: It augments `bash` to complete different sets of tasks (e.g., analysis, visualization).

Side note: If we decide to keep this `IPythonRunCellAction` as the core action, it will probably also be helpful if we pre-install the `jupyter` dependencies in the sandbox image to save the user's time from downloading.

WDYT?

Reply:
Let's keep it for now. I'm also good w/ adding jupyter to the sandbox